python/ray/worker.py

```diff
@@ -686,13 +694,69 @@ def init(
             and the API is subject to change.
 
     Returns:
-        Address information about the started processes.
+        If the provided address included a protocol (e.g. "ray://1.2.3.4")
+        then a ClientContext is returned with information such as settings,
+        server versions for ray and python, and the dashboard_url.
+        Otherwise, returns address information about the started processes.
 
     Raises:
         Exception: An exception is raised if an inappropriate combination of
             arguments is passed in.
     """
 
+    # If available, use RAY_ADDRESS to override if the address was left
+    # unspecified, or set to "auto" in the call to init
+    address_env_var = os.environ.get(
+        ray_constants.RAY_ADDRESS_ENVIRONMENT_VARIABLE)
+    if address_env_var:
+        if address is None or address == "auto":
+            address = address_env_var
+            logger.info(
+                f"Using address {address_env_var} set in the environment "
+                f"variable {ray_constants.RAY_ADDRESS_ENVIRONMENT_VARIABLE}")
+
+    if address is not None and "://" in address:
+        # Address specified a protocol, use ray client
+        builder = ray.client(address)
+        builder._store_args(
```

Comment:
Not sure how this is currently working here, but we want to use this opportunity to clean up these arguments:
- For the old way to call ray.init, all the arguments should be supported as they used to be
- For the new protocol:// convention, we should restrict the parameters that can be passed in. In particular:
For local mode, only
```
ray.init(
    address="local://",
    runtime_env={...},
    num_cpus=5,
    num_gpus=1,
    resources={...},
    dashboard_host="127.0.0.1",
    dashboard_port=8000,
    logging_format=”...”,
    logging_level=20,
    logging_to_driver=False,
    internal_config={...}
)
```
are supported. For ray:// mode, only
```
ray.init(
    address="ray://54.34.99.123:6379",
    namespace="my-namespace",
    job_name="job-0",
    runtime_env={...},
)
```
are supported.

Reply:
If any extra parameters are specified, an error should be raised. Also we should make sure only the new parameters are showing up in the documentation.

Reply:
Sounds good. Currently this works by passing everything* the user passed to init into either:
- builder._catch_all, for **kwargs
- builder._store_args, for all the other arguments

This should let us do argument validation differently for each protocol by overriding the methods in the corresponding builder (local:// [can](https://github.com/ckw017/ray/blob/9759246c8de0824e24d3d9fce96ece4e32d3b915/python/ray/tests/test_ray_init.py#L149-L151) take all the regular init args, ray:// [can't](https://github.com/ckw017/ray/blob/9759246c8de0824e24d3d9fce96ece4e32d3b915/python/ray/tests/test_ray_init.py#L171-L176))


* Currently there's a bit of a caveat: since all the init arguments come with defaults, currently figuring out which ones have changed by [comparing everything to their default value](https://github.com/ckw017/ray/blob/9759246c8de0824e24d3d9fce96ece4e32d3b915/python/ray/worker.py#L728-L743), so technically someone might do ray.init("ray://1.2.3.4", num_cpus=None) and won't get an error even though num_cpus isn't a valid argument, since num_cpus is still the default value. But case of people trying to override with default values shouldn't be too prominent or consequential.